src/singleton_snap.py

```diff
@@ -0,0 +1,293 @@
+        """
+        lock_path = self._get_registration_file_path(snap_name)
+        action = "registering" if create else "unregistering"
+
+        try:
+            with self._lock_directory():
+                if create:
+                    open(lock_path, "w").close()
+                else:
+                    try:
+                        os.remove(lock_path)
+                    except FileNotFoundError as e:
+                        # pass
+                        raise SingletonSnapError(f"Error {action} unit: {e}") from e
+        except Exception as e:
+            raise SingletonSnapError(f"Error {action} unit: {e}") from e
+
+    def get_units(self, snap_name: str) -> List[str]:
+        """Get all units currently registered for a snap (atomic with directory lock).
+
+        Args:
+            snap_name: Name of the snap to get units for
+
+        Returns:
+            List of unit names associated with the snap
+
+        Raises:
+            SingletonSnapError: If there's an error accessing the lock directory
+        """
+        prefix = f"{self.LOCK_FILE_PREFIX}{self._normalize_name(snap_name)}{self.SEPARATOR}"
+        units = []
+
+        try:
+            with self._lock_directory():
+                for filename in os.listdir(self.LOCK_DIR):
+                    if filename.startswith(prefix):
+                        units.append(filename[len(prefix) :])
+        except Exception as e:
+            raise SingletonSnapError(f"Error reading unit list: {e}") from e
+
+        return units
+
+    def is_used_by_other_units(self, snap_name: str) -> bool:
+        """Check if snap is being used by other units."""
+        return any(unit != self.unit_name for unit in self.get_units(snap_name))
+
+    @contextmanager
+    def _acquire_lock(self, lock_name: str, timeout: int) -> Generator[None, None, None]:
+        """Internal method to acquire an exclusive lock with timeout.
+
+        Uses fcntl.flock to acquire an exclusive lock on a lock file.
+        Retries until the timeout is reached.
+        """
+        lock_path = os.path.join(self.LOCK_DIR, f"{self.LOCK_FILE_PREFIX}{lock_name}")
+        with open(lock_path, "w") as f:
+            deadline = time.time() + timeout
+            while True:
+                try:
+                    fcntl.flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)
+                    break
+                except BlockingIOError:
+                    if time.time() > deadline:
+                        raise SingletonSnapError(f"Timeout acquiring lock for {lock_name}")
+                    time.sleep(0.1)
+            try:
+                yield
+            finally:
+                fcntl.flock(f, fcntl.LOCK_UN)
+                try:
+                    os.remove(lock_path)
+                except FileNotFoundError:
+                    pass
+                except OSError:
+                    raise SingletonSnapError(f"Error removing lock file for {lock_name}")
```

Comment:
Here we're throwing away the original exception. Should at least do `from e` (though per above, I'd suggest just letting the original `OSError` bubble up, not catching it at all.